src/NServiceBus.Core/Unicast/Callback.cs

```
@@ -58,60 +60,68 @@ public string MessageId
-            return Task<T>.Factory.FromAsync(asyncResult, x => completion((CompletionResult) x.AsyncState),
-                                                 TaskCreationOptions.None, TaskScheduler.Default);
+            return InternalRegister(null, null).ContinueWith(t => completion(t.Result), TaskContinuationOptions.OnlyOnRanToCompletion);
         }
 
         public Task Register(Action<CompletionResult> completion)
         {
-            var asyncResult = Register(null, null);
-            return Task.Factory.FromAsync(asyncResult, x => completion((CompletionResult) x.AsyncState),
-                                              TaskCreationOptions.None, TaskScheduler.Default);
+            return InternalRegister(null, null).ContinueWith(t => completion(t.Result), TaskContinuationOptions.OnlyOnRanToCompletion);
         }
 
         public IAsyncResult Register(AsyncCallback callback, object state)
         {
+            return InternalRegister(callback, state);
+        }
+
+        Task<CompletionResult> InternalRegister(AsyncCallback callback, object state)
+        {
             if (isSendOnly)
             {
                 throw new Exception("Callbacks are invalid in a sendonly endpoint.");
             }
-            var result = new BusAsyncResult(callback, state);
 
-            if (Registered != null)
-                Registered(this, new BusAsyncResultEventArgs { Result = result, MessageId = messageId });
+            var tcs = new TaskCompletionSource<CompletionResult>(new CompletionResult
+            {
+                State = state
+            });
+
+            if (callback != null)
+            {
+                tcs.Task.ContinueWith(t =>
+                {
+                    try
+                    {
+                        callback(t);
+                    }
+                    catch (Exception e)
+                    {
+                        log.Error(callback.ToString(), e);
```

Shouldn't we just let it bubble up ?